nexus/src/db/datastore/volume.rs

```diff
@@ -447,6 +467,195 @@ impl DataStore {
+                        volume
+                    }
+                };
+
+                // If a read_only_parent exists, remove it from volume_id, and
+                // attach it to temp_volume_id.
+                let vcr: VolumeConstructionRequest =
+                    serde_json::from_str(
+                        volume.data()
+                    )
+                    .map_err(|e| {
+                        TxnError::CustomError(
+                            RemoveReadOnlyParentError::SerdeError(
+                                e,
+                            ),
+                        )
+                    })?;
+
+                match vcr {
+                    VolumeConstructionRequest::Volume {
+                        id,
+                        block_size,
+                        sub_volumes,
+                        read_only_parent,
+                    } => {
+                        if read_only_parent.is_none() {
+                            // This volume has no read_only_parent
+                            Ok(false)
+                        } else {
+                            // Create a new VCR and fill in the contents
+                            // from what the original volume had.
+                            let new_vcr = VolumeConstructionRequest::Volume {
+                                id,
+                                block_size,
+                                sub_volumes,
+                                read_only_parent: None,
+                            };
+
+                            let new_volume_data =
+                                serde_json::to_string(
+                                    &new_vcr
+                                )
+                                .map_err(|e| {
+                                    TxnError::CustomError(
+                                        RemoveReadOnlyParentError::SerdeError(
+                                            e,
+                                        ),
+                                    )
+                                })?;
+
+                            // Update the original volume_id with the new
+                            // volume.data.
+                            use db::schema::volume::dsl as volume_dsl;
+                            let num_updated = diesel::update(volume_dsl::volume)
+                                .filter(volume_dsl::id.eq(volume_id))
+                                .set(volume_dsl::data.eq(new_volume_data))
+                                .execute(conn)?;
+
+                            // This should update just one row.  If it does
+                            // not, then something is terribly wrong in the
+                            // database.
+                            if num_updated != 1 {
+                                return Err(TxnError::CustomError(
+                                    RemoveReadOnlyParentError::UnexpectedDatabaseState,
```

Comment:
It'd be nice to have a freeform string here saying what we found (e.g., `unexpectedly updated N rows`)